### ISSUE TITLE
Nullness Checker crash: AsSuperVisitor: type is not an erased subtype of supertype 5

### ISSUE THREAD

**Issue by wmdietlGC** [typetools/checker-framework#1377]
This issue is against the code in #1373 as it is right now (including https://github.com/typetools/checker-framework/pull/1373/commits/1a8b2a4fb2771bbc4655305f6559945d78540308):

Run the Nullness Checker on:

```java
interface Function<P, R> {
  R apply(P p);
}

interface Foo<V> {
  static <U> Foo<U> of(Foo<U> in) {
    return in;
  }

  <S> Foo<S> m1(Function<? super V, S> f);
  <T> Foo<T> m2(Function<V, T> f);
}

class Crash5 {
  void foo(Foo<Void> p) {
    Foo.of(p.m1(in -> p))
        .m2(empty -> 5);
  }
}
```

This results in the following stack trace:

```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @FBCBottom @NonNull Object
  superType: @FBCBottom @NonNull Foo<@Initialized @Nullable Void>
  Compilation unit: Crash5.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:41)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:155)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:340)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:86)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:68)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:131)
  org.checkerframework.framework.type.DefaultTypeHierarchy.castedAsSuper(DefaultTypeHierarchy.java:1087)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:496)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:52)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:257)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:192)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:170)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:1814)
```

**Comment by smillst** [comment]
I can reproduce this in master, so it's not related to the fix in #1373.

**Comment by wmdietlGC** [comment]
It didn't occur with the April 2017 release, in case that helps localize the problem.
